### PR TITLE
Add support for PostgreSQL regex match

### LINKED ISSUE #795: Implement `~` pattern matching operator
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
DataFusion supports the `LIKE` operator which has a limited form of regular expression matching
It also supports `regexp_match` which adds support to extract matched strings (kudos to @seddonm1  !)
However, it doesn't expose any way to do full on posix style regular expression matching, such as `~` as described in the postgres manual: https://www.postgresql.org/docs/9.3/functions-matching.html

**Describe the solution you'd like**
Support syntax such as the following:

```sql
select * from foo where column ~ 'a.*b';
```

**Describe alternatives you've considered**
N/A

**Additional context**
@atingchen asked about this downstream in IOx: https://github.com/influxdata/influxdb_iox/issues/2137

### PR DESCRIPTION
**Pull request by b41sh**
# Which issue does this PR close?

Closes #795

 # Rationale for this change

Implement PostgreSQL pattern matching operator

# What changes are included in this PR?

1. modify sqlparser to support PostgreSQL regex match syntax [sqlparser-rs#328](https://github.com/ballista-compute/sqlparser-rs/pull/328)
2. use arrow-rs kernel `regexp_match` function to do regular expression

# Are there any user-facing changes?

support pattern matching using POSIX regular expressions

```
> CREATE EXTERNAL TABLE X STORED AS PARQUET LOCATION './benchmarks/parquet/nation/';

0 rows in set. Query took 0.005 seconds.
> select n_name from X where n_name ~ '^[A-D]';
+-----------+
| n_name    |
+-----------+
| ARGENTINA |
| BRAZIL    |
| CANADA    |
| CHINA     |
+-----------+
4 rows in set. Query took 0.010 seconds.
> select n_name from X where n_name ~* '^[a-d]';
+-----------+
| n_name    |
+-----------+
| ARGENTINA |
| BRAZIL    |
| CANADA    |
| CHINA     |
+-----------+
4 rows in set. Query took 0.008 seconds.
> select n_name from X where n_name !~ '^[C-R]';
+----------------+
| n_name         |
+----------------+
| ARGENTINA      |
| BRAZIL         |
| SAUDI ARABIA   |
| VIETNAM        |
| UNITED KINGDOM |
| UNITED STATES  |
+----------------+
6 rows in set. Query took 0.009 seconds.
> select n_name from X where n_name !~* '^[c-r]';
+----------------+
| n_name         |
+----------------+
| ARGENTINA      |
| BRAZIL         |
| SAUDI ARABIA   |
| VIETNAM        |
| UNITED KINGDOM |
| UNITED STATES  |
+----------------+
6 rows in set. Query took 0.009 seconds.
```

### CODE REVIEW

File: datafusion/src/physical_plan/expressions/binary.rs [hunk 1]
```diff
@@ -339,6 +340,63 @@ macro_rules! boolean_op {
     }};
 }
 
+/// Invoke a regexp_match kernel on a pair of arrays
+macro_rules! regexp_match_op {
+    ($LEFT:expr, $RIGHT:expr, $NOT:expr, $CASE:expr) => {{
+        match $LEFT.data_type() {
+            DataType::Utf8 => {
```
Comment (Dandandan) [inline datafusion/src/physical_plan/expressions/binary.rs:347]: `LargeUtf8`/`StringArray` would be nice to suport too I think.

Comment (b41sh) [reply datafusion/src/physical_plan/expressions/binary.rs:346]: I have added `LargeUtf8` support, PTAL @Dandandan

**Comment by alamb** [comment]
Thanks @b41sh  -- I plan to review this PR tomorrow morning. I apologize for being backed up on review

File: datafusion/src/physical_plan/expressions/binary.rs [hunk 2]
```diff
@@ -339,6 +340,72 @@ macro_rules! boolean_op {
...
+            .as_any()
+            .downcast_ref::<$ARRAYTYPE>()
+            .expect("regexp_match_array_op failed to downcast array");
+        let rr = $RIGHT
+            .as_any()
+            .downcast_ref::<$ARRAYTYPE>()
+            .expect("regexp_match_array_op failed to downcast array");
+
+        let array = match $CASE {
+            true => regexp_match(&ll, &rr, Some(&$ARRAYTYPE::from(vec!["i"; ll.len()])))?,
```
Comment (alamb) [inline datafusion/src/physical_plan/expressions/binary.rs:373]: It unfortunate that the `arrow-rs` doesn't have a `regexp_filter` style kernel that returns a BooleanArray -- 

https://sourcegraph.com/github.com/apache/arrow-rs/-/blob/arrow/src/compute/kernels/regexp.rs?L16

 I filed https://github.com/apache/arrow-rs/issues/697 and https://github.com/apache/arrow-datafusion/issues/905 to track that potential work, if we want to improve performance of these operators in future versions of DataFusion,

**Review by alamb — APPROVED** [review]
I tested this PR out locally and it looks great 🥇  Nice work @b41sh! I found the code clean and beautiful to review. 

The only other thing I can think of that might be worth adding is some end-to-end tests in `datafusion/tests/sql.rs`. 

For anyone else interested, here is how I tested:

Get updated sqlparser

```diff
diff --git a/Cargo.toml b/Cargo.toml
index d6da8c14c..bb3aa2001 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -29,3 +29,6 @@ members = [
 ]

 exclude = ["python"]
+
+[patch.crates-io]
+sqlparser = { git = "https://github.com/b41sh/sqlparser-rs.git", branch = "regexp_match"}
\ No newline at end of file
diff --git a/ballista/rust/core/Cargo.toml b/ballista/rust/core/Cargo.toml
index b2fa50c88..66f61b6e6 100644
--- a/ballista/rust/core/Cargo.toml
+++ b/ballista/rust/core/Cargo.toml
@@ -37,7 +37,7 @@ hashbrown = "0.11"
 log = "0.4"
 prost = "0.8"
 serde = {version = "1", features = ["derive"]}
-sqlparser = "0.9.0"
+sqlparser = "0.9.1-alpha.0"
 tokio = "1.0"
 tonic = "0.5"
 uuid = { version = "0.8", features = ["v4"] }
diff --git a/datafusion/Cargo.toml b/datafusion/Cargo.toml
index 286be8a7a..9819275ab 100644
--- a/datafusion/Cargo.toml
+++ b/datafusion/Cargo.toml
@@ -51,7 +51,7 @@ ahash = "0.7"
 hashbrown = { version = "0.11", features = ["raw"] }
 arrow = { version = "5.1", features = ["prettyprint"] }
 parquet = { version = "5.1", features = ["arrow"] }
-sqlparser = "0.9.0"
+sqlparser = "0.9.1-alpha.0"
 paste = "^1.0"
 num_cpus = "1.13.0"
 chrono = "0.4"
```

Run datafusion-cli:

```shell
echo "foo" > /tmp/foo.csv
echo "Bar" >> /tmp/foo.csv
echo "Baz" >> /tmp/foo.csv
echo "ZZ" >> /tmp/foo.csv
cargo run -p datafusion-cli
```

```sql
> CREATE EXTERNAL TABLE foo(a VARCHAR) STORED AS CSV LOCATION '/tmp/foo.csv';
0 rows in set. Query took 0.002 seconds.


> select * from foo where a ~ 'z';
+-----+
| a   |
+-----+
| Baz |
+-----+
1 row in set. Query took 0.012 seconds.

> select * from foo where a ~* 'z';
+-----+
| a   |
+-----+
| Baz |
| ZZ  |
+-----+
2 rows in set. Query took 0.026 seconds.

> select * from foo where a !~ 'z';
+-----+
| a   |
+-----+
| foo |
| Bar |
| ZZ  |
+-----+
3 rows in set. Query took 0.012 seconds.

> select * from foo where a !~* 'z';
+-----+
| a   |
+-----+
| foo |
| Bar |
+-----+
2 rows in set. Query took 0.012 seconds.
```

**Comment by alamb** [comment]
https://github.com/apache/arrow-datafusion/pull/934 is now merged, so upon rebase this PR should be passing

**Comment by alamb** [comment]
The python test failures have been resolved. One more rebase and I bet this one can finally be merged in. Super kudos for keeping with it @b41sh  -- thank you

**Comment by b41sh** [comment]
> The python test failures have been resolved. One more rebase and I bet this one can finally be merged in. Super kudos for keeping with it @b41sh -- thank you

I have merged the master, PTAL, thanks for your review @alamb

**Review by alamb — APPROVED** [review]
I think it looks great @b41sh  -- thank you for persevering!

**Pull request merged**